File: 백준/Silver/9012. 괄호/괄호.py
import sys

N = int(sys.stdin.readline())

lst_n = [list(sys.stdin.readline().rstrip()) for _ in range(N)]

def push(x):
    global top
    top += 1
    stack[top] = x

def popp():
    global top
    a = stack[top]
    stack[top] = 0
    top -= 1
    return a

for i in lst_n:
    stack = [0] * 51
    top = -1
    cnt = len(i)
    for j in i:
        if j == '(':
            push(j)
        elif j == ')':
            if stack[top] == '(':
                popp()
            else:
                push(j)
    if top != -1:
        print('NO')
    else:
        print('YES')

File: 백준/Silver/9012. 괄호/괄호.cc
#include <iostream>
#include <string>

using namespace std;

int cnt;

int main() {
    cin >> cnt;
    for (int i = 0; i < cnt; i++) {
        int top = -1;
        char stack[50];
        string str;
        cin >> str;
        for (int j = 0; j < str.length(); j++) {
            if (str[j] == ')') {
                if (stack[top] == '(' && top > -1) {
                    top--;
                }
                else {
                    top = 1;
                    break;
                }
            }
            else if (str[j] == '}') {
                if (stack[top] == '{' && top > -1) {
                    top--;
                }
                else {
                    top = 1;
                    break;
                }
            }
            else if (str[j] == ']') {
                if (stack[top] == '[' && top > -1) {
                    top--;
                }
                else {
                    top = 1;
                    break;
                }
            }
            else {
                top++;
                stack[top] = str[j];
            }
        }
        if (top == -1) {
            cout << "YES" << '\n';
        }
        else {
            cout << "NO" << '\n';
        }
    }

    return 0;
}